File: bot_template/bot.py
#!/usr/bin/env python3

import json
import sys


def build_opponent_stats(state: dict) -> dict[str, dict[str, int]]:
    hero_id = state["hero"]["player_id"]
    stats: dict[str, dict[str, int]] = {}
    for event in state["action_history"]:
        player_id = event["player_id"]
        if player_id == hero_id:
            continue
        player_stats = stats.setdefault(player_id, {"actions": 0, "aggressive_actions": 0})
        player_stats["actions"] += 1
        if event["action"] in {"bet", "raise"}:
            player_stats["aggressive_actions"] += 1
    return stats


def choose_action(state: dict) -> dict:
    hero = state["hero"]
    to_call = hero["to_call"]
    pot = state["board"]["pot"]
    legal = {entry["action"]: entry for entry in state["legal_actions"]}

    opponent_stats = build_opponent_stats(state)
    aggressive_opponents = sum(
        1
        for values in opponent_stats.values()
        if values["aggressive_actions"] * 2 >= values["actions"] and values["actions"] > 0
    )

    if "check" in legal:
        return {"action": "check"}

    if "call" in legal and aggressive_opponents == 0 and to_call <= pot // 4:
        return {"action": "call"}

    if "fold" in legal:
        return {"action": "fold"}

    first = next(iter(legal.values()))
    response = {"action": first["action"]}
    if "min_amount" in first:
        response["amount"] = first["min_amount"]
    return response


def main() -> int:
    state = json.load(sys.stdin)
    response = choose_action(state)
    json.dump(response, sys.stdout, separators=(",", ":"))
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: bot/examples/go_bot/bot.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

type LegalAction struct {
	Action string `json:"action"`
}

type State struct {
	LegalActions []LegalAction `json:"legal_actions"`
}

type Response struct {
	Action string `json:"action"`
}

func main() {
	var state State
	if err := json.NewDecoder(os.Stdin).Decode(&state); err != nil {
		os.Exit(1)
	}

	response := Response{Action: "fold"}
	for _, entry := range state.LegalActions {
		if entry.Action == "check" {
			response.Action = "check"
			break
		}
		if entry.Action == "call" {
			response.Action = "call"
		}
	}

	if err := json.NewEncoder(os.Stdout).Encode(response); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}
